File: results/0.5/DNS/dns-gpt-4o.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <hammer/hammer.h>

typedef struct {
    uint16_t id;
    uint16_t flags;
    uint16_t qdcount;
    uint16_t ancount;
    uint16_t nscount;
    uint16_t arcount;
} dns_header_t;

typedef struct {
    char *qname;
    uint16_t qtype;
    uint16_t qclass;
} dns_question_t;

typedef struct {
    char *name;
    uint16_t type;
    uint16_t class;
    uint32_t ttl;
    uint16_t rdlength;
    uint8_t *rdata;
} dns_rr_t;

HParser *dns_header_parser;
HParser *dns_question_parser;
HParser *dns_rr_parser;
HParser *dns_message_parser;

HParser *create_dns_header_parser() {
    return h_sequence(
        h_uint16(), // ID
        h_uint16(), // Flags
        h_uint16(), // QDCOUNT
        h_uint16(), // ANCOUNT
        h_uint16(), // NSCOUNT
        h_uint16(), // ARCOUNT
        NULL
    );
}

HParser *create_dns_question_parser() {
    HParser *label = h_length_value(h_uint8(), h_repeat_n(h_uint8(), 0));
    HParser *qname = h_many1(label);
    return h_sequence(
        qname,      // QNAME
        h_uint16(), // QTYPE
        h_uint16(), // QCLASS
        NULL
    );
}

HParser *create_dns_rr_parser() {
    HParser *label = h_length_value(h_uint8(), h_repeat_n(h_uint8(), 0));
    HParser *name = h_many1(label);
    return h_sequence(
        name,       // NAME
        h_uint16(), // TYPE
        h_uint16(), // CLASS
        h_uint32(), // TTL
        h_length_value(h_uint16(), h_repeat_n(h_uint8(), 0)), // RDLENGTH + RDATA
        NULL
    );
}

HParser *create_dns_message_parser() {
    HParser *questions = h_many(create_dns_question_parser());
    HParser *answers = h_many(create_dns_rr_parser());
    HParser *authorities = h_many(create_dns_rr_parser());
    HParser *additionals = h_many(create_dns_rr_parser());

    return h_sequence(
        create_dns_header_parser(),
        questions, // QDCOUNT
        answers,   // ANCOUNT
        authorities, // NSCOUNT
        additionals, // ARCOUNT
        NULL
    );
}

void parse_dns_message(const char *filename) {
    FILE *file = fopen(filename, "rb");
    if (!file) {
        perror("Failed to open file");
        exit(EXIT_FAILURE);
    }

    fseek(file, 0, SEEK_END);
    long length = ftell(file);
    fseek(file, 0, SEEK_SET);

    uint8_t *buffer = malloc(length);
    if (!buffer) {
        perror("Failed to allocate memory");
        fclose(file);
        exit(EXIT_FAILURE);
    }

    fread(buffer, 1, length, file);
    fclose(file);

    HParseResult *result = h_parse(dns_message_parser, buffer, length);
    if (result) {
        printf("DNS message parsed successfully.\n");
        h_parse_result_free(result);
    } else {
        printf("Failed to parse DNS message.\n");
    }

    free(buffer);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <dns_message_file>\n", argv[0]);
        return EXIT_FAILURE;
    }

    dns_message_parser = create_dns_message_parser();
    parse_dns_message(argv[1]);
    h_parser_free(dns_message_parser);

    return EXIT_SUCCESS;
}

File: results/0.5/DNS/dns-gpt-4o.rs
use nom::{
    bytes::complete::{tag, take},
    combinator::map_res,
    multi::count,
    number::complete::{be_u16, be_u32, be_u8},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
struct DnsHeader {
    id: u16,
    flags: u16,
    qdcount: u16,
    ancount: u16,
    nscount: u16,
    arcount: u16,
}

#[derive(Debug)]
struct DnsQuestion {
    qname: Vec<String>,
    qtype: u16,
    qclass: u16,
}

#[derive(Debug)]
struct DnsRecord {
    name: Vec<String>,
    rtype: u16,
    rclass: u16,
    ttl: u32,
    rdlength: u16,
    rdata: Vec<u8>,
}

#[derive(Debug)]
struct DnsMessage {
    header: DnsHeader,
    questions: Vec<DnsQuestion>,
    answers: Vec<DnsRecord>,
    authorities: Vec<DnsRecord>,
    additionals: Vec<DnsRecord>,
}

fn parse_dns_header(input: &[u8]) -> IResult<&[u8], DnsHeader> {
    let (input, id) = be_u16(input)?;
    let (input, flags) = be_u16(input)?;
    let (input, qdcount) = be_u16(input)?;
    let (input, ancount) = be_u16(input)?;
    let (input, nscount) = be_u16(input)?;
    let (input, arcount) = be_u16(input)?;
    Ok((
        input,
        DnsHeader {
            id,
            flags,
            qdcount,
            ancount,
            nscount,
            arcount,
        },
    ))
}

fn parse_labels(input: &[u8]) -> IResult<&[u8], Vec<String>> {
    let mut labels = Vec::new();
    let mut input = input;
    loop {
        let (i, length) = be_u8(input)?;
        if length == 0 {
            break;
        }
        let (i, label) = take(length)(i)?;
        labels.push(String::from_utf8_lossy(label).into_owned());
        input = i;
    }
    Ok((input, labels))
}

fn parse_question(input: &[u8]) -> IResult<&[u8], DnsQuestion> {
    let (input, qname) = parse_labels(input)?;
    let (input, qtype) = be_u16(input)?;
    let (input, qclass) = be_u16(input)?;
    Ok((input, DnsQuestion { qname, qtype, qclass }))
}

fn parse_record(input: &[u8]) -> IResult<&[u8], DnsRecord> {
    let (input, name) = parse_labels(input)?;
    let (input, rtype) = be_u16(input)?;
    let (input, rclass) = be_u16(input)?;
    let (input, ttl) = be_u32(input)?;
    let (input, rdlength) = be_u16(input)?;
    let (input, rdata) = take(rdlength)(input)?;
    Ok((
        input,
        DnsRecord {
            name,
            rtype,
            rclass,
            ttl,
            rdlength,
            rdata: rdata.to_vec(),
        },
    ))
}

fn parse_dns_message(input: &[u8]) -> IResult<&[u8], DnsMessage> {
    let (input, header) = parse_dns_header(input)?;
    let (input, questions) = count(parse_question, header.qdcount as usize)(input)?;
    let (input, answers) = count(parse_record, header.ancount as usize)(input)?;
    let (input, authorities) = count(parse_record, header.nscount as usize)(input)?;
    let (input, additionals) = count(parse_record, header.arcount as usize)(input)?;
    Ok((
        input,
        DnsMessage {
            header,
            questions,
            answers,
            authorities,
            additionals,
        },
    ))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <dns_message_file>", args[0]);
        return;
    }

    let filename = &args[1];
    let mut file = File::open(filename).expect("Failed to open file");
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("Failed to read file");

    match parse_dns_message(&buffer) {
        Ok((_, message)) => println!("{:?}", message),
        Err(e) => eprintln!("Failed to parse DNS message: {:?}", e),
    }
}